firedrake/assemble.py

```diff
@@ -214,11 +216,23 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
         kernels = tsfc_interface.compile_form(f, "form", parameters=form_compiler_parameters, inverse=inverse)
         integral_types = [integral.integral_type() for integral in f.integrals()]
 
+        if bcs is not None:
+            for bc in bcs:
+                if isinstance(bc, FormBC):
+                    integral_types += [integral.integral_type() for integral in bc.f.integrals()]
     rank = len(f.arguments())
 
     is_mat = rank == 2
     is_vec = rank == 1
 
+    # FormBCs requre recursive calls of _assemble(...), and kernels keep
+    # being appended to a sigle list ("loops") at all levels of recursion.
+    # The first level call of _assemble(...) is identified by "loops is None."
+    first_call = False
+    if loops is None:
+        loops = []
+        first_call = True
```

Comment:
I think we should think about how we can refactor this loop collection so that it is less messy.

Thinking about it, since it is an internal implementation detail, we could always be in "collecting loops mode" in here. And then the top-level `assemble` could collect the loops and immediately evaluate them.

We could then make this function `yield` loops (i.e. become a generator) and the recursive call to assemble collecting the loops can be replaced by `yield from _assemble(...)`

Reply:
This would also simplify the code below, because the `thunk` call would just become the list of loops we need to evaluate (I think...)

Reply:
Thanks! I will make changes based on this suggestion.